### ISSUE TITLE
Reason for Property and Node to be enums?

### ISSUE THREAD

**Issue by Furao** [yodaldevoid/dts_viewer#13]
I am using the `device_tree_source` library and I was wondering what your reasons were for making Property and Node enums? 

My use case is to parse the device tree to create some structs of my own to be consumed by another tool. I am finding it fairly difficult to keep parsing these nested enums while looking for specific information.

Consider a function where I expect a node to have a property where the value is a single string:

```rust
fn get_node_prop_string(n: &Node, prop: &str) -> Option<String> {
     match n {
        Node::Existing { proplist, .. } => 
        {
            match proplist.get(prop)? {
                Property::Existing { val, .. } => {
                    val.as_ref()?.iter().next().map(|z| match z {
                        Data::String(s) => Some(s.clone()),
                        _ => None,
                    }).flatten()
                },
                Property::Deleted { .. } => None,
            }
        },
        Node::Deleted { name, .. } => None,
     }
}
```
I need to have 3 matches in there, because  there are 3 nested enums. I can make it cleaner with some other helper functions:
```rust
fn get_node_proplist(n: &Node) -> Option<&HashMap<String, Property>> {
    match n {
        Node::Existing { proplist, .. } => Some(proplist),
        Node::Deleted { .. } => None,
    }
}

fn get_property_val(p: &Property) -> &Option<Vec<Data>> {
    match p {
        Property::Existing { val, .. } => val,
        Property::Deleted { .. } => &None,
    }
}

fn get_node_prop_string(n: &Node, prop: &str) -> Option<String> {
    let p = get_node_proplist(n)?.get(prop)?;
    let vals = get_property_val(p).as_ref()?;
    match &vals[0] {
        Data::String(s) => Some(s.clone()),
        _ => None,
    }
}
```
But I need to write those functions myself. 

Here is an alternative approach:

```rust
#[derive(PartialEq, Eq, Debug, Clone)]
pub struct PropertyInner {
    val: Option<Vec<Data>>,
    labels: Vec<String>,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Property {
    name: String,
    offset: usize,
    exists: Option<PropertyInner>,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct NodeInner {
    proplist: HashMap<String, Property>,
    children: HashMap<String, Node>,
    labels: Vec<String>,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Node {
    name: NodeName,
    offset: usize,
    exists: Option<NodeInner>,
}
```
Then my function becomes a lot cleaner by default:
```rust
fn node_get_prop_string_by_name(node: &Node, prop: &str) -> Option<String> {
    let vals = node.exists.as_ref()?
        .proplist.get(prop)?
        .exists.as_ref()?
        .val.clone()?;
    
    match &vals[0] {
        Data::String(s) => Some(s.clone()),
        _ => None,
    }
}
```
This simplifies my use case, but also still lets you tell if a node or property is deleted. You just need to check the `exists` option. 

I am fairly new to rust so I may be missing something here.

**Comment by yodaldevoid** [comment]
This project was my first one in Rust and I haven't touched it in nearly 6 years.

I don't currently work with device trees in any capacity so I don't really plan on working on this at the moment.

As for why I organized it this way, it's been long enough that I don't really remember, but I have to assume I was far too enamored with Rust's enums and so I used them event where things could have been done better a different way.
